### PR TITLE
[CA-849] Trace auth calls

### PR DESCRIPTION
**Pull request by andy7i**
Ticket: https://broadworkbench.atlassian.net/browse/CA-849
Goal: trace auth calls.
Follow on work from Sam tracing work: https://github.com/broadinstitute/sam/pull/421#discussion_r422182607

<Put notes here to help reviewer understand this PR>

This PR includes:
Adds tracing for auth calls
Adds SamRequestContextDirectives
Adds a new directive `withNewTraceSpan` 
Light plumbing

Example trace:
Screen Shot 2020-05-27 at 7 01 27 PM

---

**PR checklist**

- [x] I've followed [the instructions](https://github.com/broadinstitute/sam/blob/develop/CONTRIBUTING.md#api-changes) if I've made any changes to the API, _especially_ if they're breaking changes
- [x] I've updated the RC_XXX release ticket with any manual steps required to release this change
- [x] I've updated the [FISMA documentation](https://github.com/broadinstitute/sam/blob/develop/CONTRIBUTING.md#fisma-documentation-changes) if I've made any security-related changes, including auth, encryption, or auditing

### CODE REVIEW

**Comment by andy7i** [comment]
jenkins retest -- fiab-start failed with an unrelated issue
```
10:46:02 [ERROR] POST https://us-central1-broad-workbench-tos-qa.cloudfunctions.net/tos/v1/user/response failed with status 400
10:46:02 "Error writing user response: TermsOfService fiab-automation-upward-dragon/1 does not exist."
```

**Comment by andy7i** [comment]
jenkins retest - fiab-start failed with:
```
10:58:34 ERROR: for consent-mongo-1  sandbox b00909f34cd031047318b0de30de5d0c8080968113c5f4f3612dd88fc4972a37 not found
10:58:34 
10:58:34 ERROR: for agora-mongo  sandbox 254229972b9b9b4f520468e19a0035484e3cdb2d74f41f09df7e5f83af4964f5 not found
10:58:34 
10:58:34 ERROR: for bond-app  sandbox 3ba27453366fe5691bb083878e45d6fb6a385a2ca6b4ab0c8c2b008d395e5e84 not found

```

**Comment by andy7i** [comment]
jenkins retest - fiab-start failed with:
```
12:26:00 ERROR: for consent-mongo-1  sandbox b00909f34cd031047318b0de30de5d0c8080968113c5f4f3612dd88fc4972a37 not found
12:26:00 
12:26:00 ERROR: for agora-mongo  sandbox 254229972b9b9b4f520468e19a0035484e3cdb2d74f41f09df7e5f83af4964f5 not found
12:26:00 
12:26:00 ERROR: for bond-app  sandbox 3ba27453366fe5691bb083878e45d6fb6a385a2ca6b4ab0c8c2b008d395e5e84 not found
```

**Comment by andy7i** [comment]
jenkins retest - leo test aborted after being stuck for 2 hours:
```
docker ps (before kill running containers older than 2 days):
CONTAINER ID        IMAGE                                                          COMMAND                  CREATED             STATUS                          PORTS               NAMES
13ceb7adc504        gcr.io/broad-dsp-gcr-public/leonardo-cluster-cleanup:initial   "/bin/sh"                2 days ago          Up 2 days                                           sql-client-container
4192782a9dae        owasp/zap2docker-weekly                                        "zap-baseline.py -t …"   3 weeks ago         Up 3 weeks (health: starting)                       friendly_einstein
Build timed out (after 120 minutes). Marking the build as aborted.
```

**Comment by andy7i** [comment]
jenkins retest - rawls tests passed but then timed out
```
Tests Passed

Error response from daemon: No such container: automation-s2gyocwy
Test container already stopped
Build timed out (after 120 minutes). Marking the build as aborted.
```
https://fc-jenkins.dsp-techops.broadinstitute.org/job/rawls-fiab-test-runner/22496/console

**Comment by andy7i** [comment]
jenkins retest - quay issues

**Review by gpolumbo-broad — APPROVED** [review]

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/api/SamRequestContextDirectives.scala [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+  /**
+    * Provides a new SamRequestContext with a tracing span that is a child of the existing SamRequestContext's
+    * `parentSpan`. If a parentSpan does not exist, this will NOT create a new one.
+    *
+    * @param spanName name of the new parentSpan in the new SamRequestContext. This span is a child of the existing
+    *                 parentSpan.
+    * @param samRequestContext the existing samRequestContext.
+    */
+  def withNewTraceSpan(spanName: String, samRequestContext: SamRequestContext): Directive1[SamRequestContext] =
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/api/SamRequestContextDirectives.scala]: I would call this `withChildTraceSpan`

**Review by dvoet — APPROVED** [review]
naming nit then 👍

**Pull request merged**